plugins/range/plugin.go

```diff
@@ -72,6 +72,23 @@ func (p *PluginState) Handler4(req, resp *dhcpv4.DHCPv4) (*dhcpv4.DHCPv4, bool)
 		}
 		p.Recordsv4[req.ClientHWAddr.String()] = &rec
 		record = &rec
+	} else if req.MessageType() == dhcpv4.MessageTypeRelease {
+		// Release the IP address from allocator
+		if freeErr := p.allocator.Free(net.IPNet{IP: record.IP}); freeErr != nil {
+			log.Errorf("Could not free IP %s for MAC %s: %v", record.IP.String(), req.ClientHWAddr.String(), freeErr)
+			return resp, false
+		}
```

Comment:
The order of freeing is backwards imo:
* If freeing from the allocator succeeds but deleting from the lease storage fails, we can end up re-assigning the IP to another client despite it being in the lease here, and then when this client comes back we give it back its previous record with the (now-reallocated) IP
* If deleting from the lease storage happens first and fails, then we're in the same situation as if no DHCPRELEASE was sent
* If deleting from the lease storage happens first, succeeds, then freeing from the allocator fails, we leak an IP; unfortunate but less incorrect than assigning the same IP to 2 different hosts